lib/websocket_server.rb

```diff
@@ -1,56 +1,67 @@
+#
+# It uses socket hijacking to retrieve the underlying TCP socket connection
+# of the HTTP request incoming from a client. This socket is detached from
+# the Rack middleware and handled by a separate transmitter thread.
+# The remote address of the console is determined by a one time secret that
+# is part of the URL and it points to a SystemConsole record in the database.
+#
+# On the lowest level the transmitter thread operates with regular sockets,
+# however, both on the client and server side wrappers are used to translate
+# between a pair of sockets. Due to some limitations in the websocket driver
+# gem, the reading/writing operations are defined in the fetch/issue methods.
+#
+# The selector for the sockets is provided by an external gem and it handles
+# the dependency between the socket pairs, i.e. one should be ready to read
+# while other one should be ready to write at the same time. The pairs ready
+# for transmission are handled by the `each_ready` iterator. As the iterator
+# always returns with a socket to read and a socket to write, the `@adapters`
+# hash has been used to access the corresponding wrappers.
+
 class WebsocketServer
   attr_accessor :logger
 
-  Pairing = Struct.new(:is_ws, :proxy)
+  RACK_404 = [404, {'Content-Type' => 'text/plain'}, ['Not found']].freeze
+  RACK_YAY = [-1, {}, []].freeze
 
   def initialize(options = {})
     @logger = options.fetch(:logger, $websocket_log)
-    logger.info('Initializing websocket worker!')
-    @pairing = {}
-    @sockets = Concurrent::Array.new
+    @logger.info('Initializing websocket server!')
+    @proxy = SurroGate.new(logger)
+    @adapters = {}
 
     @transmitter = Thread.new do
       loop do
-        begin
-          reads, writes, errors = IO.select(@sockets, @sockets, @sockets, 1)
-        rescue IOError
-          @sockets.select(&:closed?).each { |err| cleanup(err) }
-        else
-          Array(errors).each { |err| cleanup(err) }
-        end
-
-        # Skip this loop if we can't do anything
-        if Array(reads).empty? || writes.empty? || errors.any?
-          sleep(1) if @sockets.empty?
-          next
-        end
+        @proxy.select(1000)
 
-        # Do the data transfers
-        reads.each do |socket|
+        @proxy.each_ready do |left, right|
           begin
-            @pairing[socket].proxy.transmit(writes, @pairing[socket].is_ws)
-          rescue => error
-            cleanup(socket, error)
+            @adapters[left].fetch(64.kilobytes) { |data| @adapters[right].issue(data) } # left -> right
+          rescue IOError, EOFError
+            cleanup(:info, "Closing websocket proxy for VM %{vm_id}", left, right)
+          rescue StandardError => ex
+            cleanup(:error, "Websocket proxy for VM %{vm_id} errored with #{ex} #{ex.backtrace.join("\n")}", left, right)
           end
         end
       end
     end
+
+    @transmitter.abort_on_exception = true
```

Comment:
Threads are hard.  😨

Reply:
Dr. Skateman or: How I Learned to Stop Worrying and Love the Threads :rofl:

Reply:
I gotta say, whenever I see `abort_on_exception = true`, I have to ask, how many times did it silently fail in a thread before you added this?  If it's less than 3, you're amazing.  The gut feeling when doing threads is "oh, I won't need to worry about exceptions raised in the thread, it's such simple work..."  :trollface:

Reply:
You don't want to know :trollface: